docs/design2.md

```diff
@@ -0,0 +1,102 @@
+|--------------|---------------------|--------------|
+| hr           | Tractable Bandwidth | Unusable     |
+| min          | Tractable Bandwidth | Opt-in       |
+| s            |                     | Tractable    |
+
+
+## Data Structures
+
+### Gossip Network
+The job of the gossip network is to ensure that there is a single, globally consistent lineage of log roots. 
+To keep $n^2$ communication costs low, and to prevent sybil attacks, clients use a small ~20 set of gossip nodes.
+
+Each gossip node fetches the latest signed log root (SLRs) and verifies consistency with all previously seen roots.
+After verifying, the gossip node signs the log root.
+
+Gossip nodes offer the following APIs:
+
+1. Get Signed Root 
+
+### Log of Map Roots
+
+The log server collects signatures from the gossip nodes to produce a Signed Log Root that has a quorum of gossip signatures.
+
+The log server contains a list of map roots representing sequential snapshots of a key-value directory.
+
+Log servers offer the following APIs:
+
+1. Latest Signed Log Root
+1. Get consistency proof between log size a and b. 
+1. Get log item i with inclusion proof to log size a.
+
+### Map Root Snapshots 
+Each Map Root represents a snapshot of a key-value dictionary.
+The map is implemented as a sparse merkle tree of fixed depth. This should be changed to a patricia-prefix-tri for better efficiency.
+
+* Indexes in the map are randomized and privacy protected via a Verifiable Random Function.
+* Values in the map represent the full history of values that have ever been stored at this index. This is accomplished by storing the merkle root of a mini-log of these values.
+  
+The map offers the following APIs:
+
+1. Get map value at index j with inclusion proof to snapshot a.
+
+### Value History Log
+These mini logs store not just the latest value, but also store every previous value in order.
+
+These mini logs are what users query when looking up their own key history, and they are what their peers verify in order to ensure that no history has been lost. 
+
+The value log offers the following API:
+
+1. Get latest value at snapshot z with inclusion proof.
+2. Get consistency proof between snapshot roots y and z.
+3. Get range of historical values between snaptots y and z.
+
+## Client Verification 
+
+Key Transparency clients store 
+1. The root of the log of map roots. This ensures that the client is using the same snapshots as the rest of the world.
+1. The root of every mini-log they have queried. This ensures that snapshots represent append-only representations of the world.
+
+When querying, Key Transparency clients ask for proof that the current snapshot and value are consistent with previous values that the client has observed.
+
+## Efficiency Improvements
+
+Generating append-only proofs for large sparse merkle trees is not efficient. For N changes per snapshot in a map of size M over T snapshots is roughly O(T * N log M) 
```

Comment:
O(...) operations/hashes?

Feels like there's a sentence missing from the end here: "This can be improved, by:" or similar?

Reply:
Changed to *Efficiency Innovations*